sycl/include/sycl/ext/oneapi/matrix/matrix-jit-use.hpp

```diff
@@ -0,0 +1,669 @@
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+namespace matrix {
+
+// packed_a and packed_b will be replaced by packed once the use implementation
+// is stable.
+enum class matrix_layout { row_major, col_major, packed_a, packed_b, unused };
+
+template <matrix_layout Layout> struct spv_matrix_layout_traits {
+  static constexpr __spv::MatrixLayout value = __spv::MatrixLayout::Unused;
+};
+
+#define SPV_MATRIX_LAYOUT_TRAITS(LAYOUT, SPV_LAYOUT)                           \
+  template <> struct spv_matrix_layout_traits<LAYOUT> {                        \
+    static constexpr __spv::MatrixLayout value = SPV_LAYOUT;                   \
+  };
+
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::row_major,
+                         __spv::MatrixLayout::RowMajor)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::col_major,
+                         __spv::MatrixLayout::ColumnMajor)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::packed_a, __spv::MatrixLayout::PackedA)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::packed_b, __spv::MatrixLayout::PackedB)
+SPV_MATRIX_LAYOUT_TRAITS(matrix_layout::unused, __spv::MatrixLayout::Unused)
+
+// unnecessary was introduced for backward compatibility.
+// Once the use implementation is stable, "unnecessary" value will be omitted
+enum class matrix_use { matrix_a, matrix_b, accumulator, unnecessary };
+
+template <matrix_use Use> struct spv_matrix_use_traits {
+  static constexpr __spv::MatrixUse value = __spv::MatrixUse::MatrixA;
+};
+
+#define SPV_MATRIX_USE_TRAITS(USE, SPV_USE)                                    \
+  template <> struct spv_matrix_use_traits<USE> {                              \
+    static constexpr __spv::MatrixUse value = SPV_USE;                         \
+  };
+
+SPV_MATRIX_USE_TRAITS(matrix_use::matrix_a, __spv::MatrixUse::MatrixA)
+SPV_MATRIX_USE_TRAITS(matrix_use::matrix_b, __spv::MatrixUse::MatrixB)
+SPV_MATRIX_USE_TRAITS(matrix_use::accumulator, __spv::MatrixUse::Accumulator)
+SPV_MATRIX_USE_TRAITS(matrix_use::unnecessary, __spv::MatrixUse::Unnecessary)
+
+template <typename G> struct spv_scope_traits {};
+template <> struct spv_scope_traits<sycl::sub_group> {
+  constexpr static auto value = __spv::Scope::Subgroup;
+};
+template <int D> struct spv_scope_traits<sycl::group<D>> {
+  constexpr static auto value = __spv::Scope::Workgroup;
+};
+
+template <typename T, size_t NumRows, size_t NumCols, matrix_use Use,
+          matrix_layout Layout = matrix_layout::unused,
+          typename Group = sycl::sub_group>
+class wi_data;
+template <typename T, size_t NumRows, size_t NumCols, matrix_use Use,
+          matrix_layout Layout = matrix_layout::unused,
+          typename Group = sycl::sub_group>
+struct joint_matrix {
+public:
+  __spv::__spirv_JointMatrixINTEL<
+      T, NumRows, NumCols, spv_matrix_use_traits<Use>::value,
+      spv_matrix_layout_traits<Layout>::value> *spvm;
+  joint_matrix(Group sg) {
+#ifndef __SYCL_DEVICE_ONLY__
+    (void)sg;
```

Comment:
```suggestion
  joint_matrix(Group) {
#ifndef __SYCL_DEVICE_ONLY__
```

Reply:
Maybe let's keep the unused "sg" in joint_matrix_fill to match the other joint matrix function?

Reply:
Ok, no objections to that, but we should probably move `(void)sg;` outside of `#ifndef`, because it is unused outside of it as well and will generate a warning in device code